tct/src/epoch.rs

```
@@ -94,35 +165,56 @@ impl Epoch {
+    /// Returns [`InsertError`] if any of:
+    ///
+    /// - the [`Epoch`] is full, or
+    /// - the most recent [`Block`] is full or was inserted by
+    ///   [`insert_block_root`](Epoch::insert_block_root).
+    pub fn insert(
+        &mut self,
+        witness: Witness,
+        commitment: impl Into<Commitment>,
+    ) -> Result<Position, InsertError> {
+        // The position at which we will insert the commitment
+        let position = self.position();
+
         self.as_mut()
             .insert(match witness {
-                Keep => Insert::Keep(commitment),
-                Forget => Insert::Hash(Hash::of(commitment)),
+                Keep => Insert::Keep(commitment.into()),
+                Forget => Insert::Hash(Hash::of(commitment.into())),
             })
             .map(|replaced| {
                 // When inserting into an epoch that's not part of a larger eternity, we should never return
                 // further indices to be forgotten, because they should be forgotten internally
                 debug_assert!(replaced.is_none());
-            })
+            })?;
+
+        Ok(position)
     }
 
     /// Get a [`Proof`] of inclusion for the commitment at this index in the epoch.
     ///
     /// If the index is not witnessed in this epoch, return `None`.
-    pub fn witness(&self, commitment: Commitment) -> Option<Proof> {
+    pub fn witness(&self, commitment: impl Into<Commitment>) -> Option<Proof> {
+        let commitment = commitment.into();
+
         let index = *self.index.get(&commitment)?;
 
-        let (auth_path, leaf) = self.inner.witness(index)?;
+        let (auth_path, leaf) = match self.inner.witness(index) {
+            Some(witness) => witness,
+            None => panic!(
+                "commitment `{:?}` indexed with position `{:?}` must be witnessed",
+                commitment, index
+            ),
+        };
```

Why are we panicking here instead of just returning `None` as before?